src/net/neighbors/db.rs

```diff
@@ -0,0 +1,652 @@
+        conn: &DBConn,
+        nk: &NeighborKey,
+    ) -> Result<Option<u32>, net_error> {
+        let mut slots = PeerDB::peer_slots(conn, nk.network_id, &nk.addrbytes, nk.port)
+            .map_err(net_error::DBError)?;
+
+        if slots.len() == 0 {
+            // not present
+            return Ok(None);
+        }
+
+        let mut rng = thread_rng();
+        slots.shuffle(&mut rng);
+
+        for slot in slots {
+            let peer_opt =
+                PeerDB::get_peer_at(conn, nk.network_id, slot).map_err(net_error::DBError)?;
+
+            match peer_opt {
+                None => {
+                    continue;
+                }
+                Some(_) => {
+                    return Ok(Some(slot));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// implements add_or_schedule_replace_neighbor
+    fn inner_add_or_schedule_replace_neighbor(
+        network: &mut PeerNetwork,
+        naddr: &NeighborAddress,
+        preamble: &Preamble,
+        handshake: &HandshakeData,
+        db_data: Option<&StackerDBHandshakeData>,
+        replacements: &mut NeighborReplacements,
+    ) -> Result<(bool, Neighbor), net_error> {
+        let block_height = network.get_chain_view().burn_block_height;
+        let mut tx = network.peerdb_tx_begin()?;
+        let mut neighbor_from_handshake = Neighbor::from_handshake(
+            &mut tx,
+            preamble.peer_version,
+            preamble.network_id,
+            handshake,
+        )?;
+
+        if Neighbor::from_neighbor_address(&mut tx, preamble.network_id, block_height, naddr)?
+            .is_some()
+        {
+            test_debug!("already know about {:?}", naddr);
+            neighbor_from_handshake
+                .save_update(&mut tx, db_data.map(|x| x.smart_contracts.as_slice()))?;
+            tx.commit()?;
+
+            // seen this neighbor before
+            return Ok((false, neighbor_from_handshake));
+        }
+
+        debug!("new neighbor {:?}", &neighbor_from_handshake.addr);
+
+        // didn't know about this neighbor yet. Try to add it.
+        let added =
+            neighbor_from_handshake.save(&mut tx, db_data.map(|x| x.smart_contracts.as_slice()))?;
+
+        if added {
+            // neighbor was new, and we had space to add it.
+            tx.commit()?;
+            return Ok((true, neighbor_from_handshake));
+        }
+
+        // neighbor was new, but we don't have space to insert it.
+        // find and record a neighbor it would replace.
+        let replaced_neighbor_slot_opt =
+            Self::find_replaced_neighbor_slot(&mut tx, &neighbor_from_handshake.addr)?;
+        if let Some(slot) = replaced_neighbor_slot_opt {
+            // if this peer isn't allowed or denied, then consider
+            // replacing.  Otherwise, keep the local configuration's preference.
+            if !neighbor_from_handshake.is_denied() && !neighbor_from_handshake.is_allowed() {
```

Comment:
Why should an allowed peer not get added?

Reply:
We don't want to have the neighbor walk algorithm override local policies regarding whether or not a node is set as allowed by the operator, or set as denied by the operator.  In either case, the peer's record needs to stick around in the DB.